Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/fragment/SubmissionQRCodeScanFragment.kt

```diff
@@ -123,10 +126,21 @@ class SubmissionQRCodeScanFragment : Fragment(R.layout.fragment_submission_qr_co
             }
 
             if (ApiRequestState.SUCCESS == it) {
-                doNavigate(
-                    SubmissionQRCodeScanFragmentDirections
-                        .actionSubmissionQRCodeScanFragmentToSubmissionResultFragment()
-                )
+                submissionViewModel.deviceUiState.observe2(this) { uiState ->
+                    SubmissionRepository.refreshDeviceUIState(true)
+                    if (uiState != DeviceUIState.UNPAIRED) {
+                        if (uiState == DeviceUIState.PAIRED_REDEEMED) {
+                            showRedeemedTokenWarningDialog()
+                            submissionViewModel.deregisterTestFromDevice()
+                            goBack()
+                        } else {
+                            doNavigate(
+                                SubmissionQRCodeScanFragmentDirections
+                                    .actionSubmissionQRCodeScanFragmentToSubmissionResultFragment()
+                            )
+                        }
```

Comment:
in my opinion this logic should be moved to the view model, then send events for navigation to fragment via life data

Reply:
Hi Chris,

the logic has been revamped. Please take a look again. and let me know if you have any question.